src/System.Text.Json/src/System/Text/Json/CustomUncheckedBitArray.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    // Inspired by BitArray
+    // https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/BitArray.cs
+    internal sealed class CustomUncheckedBitArray
+    {
+        private int[] _array;
+
+        public int Length { get; private set; }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public CustomUncheckedBitArray(int bitLength)
+        {
+            Debug.Assert(bitLength > 0 && bitLength % 32 == 0, $"bitLength: {bitLength}");
+
+            _array = new int[bitLength / 32];
+            Length = bitLength;
+        }
+
+        public bool this[int index]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                Debug.Assert(index >= 0 && index < Length, $"index: {index}, Length: {Length}");
+
+                int elementIndex = Div32Rem(index, out int extraBits);
+
+                Debug.Assert(elementIndex < _array.Length, $"index: {index}, Length: {Length}, elementIndex: {elementIndex}, arrayLength: {_array.Length}, extraBits: {extraBits}");
+
+                return (_array[elementIndex] & (1 << extraBits)) != 0;
+            }
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            set
+            {
+                Debug.Assert(index >= 0, $"index: {index}, Length: {Length}");
+
+                // Grow the array when setting a bit if it isn't big enough
+                // This way the caller doesn't have to check.
+                if (index >= Length)
+                {
+                    Grow(index);
+                }
+
+                int elementIndex = Div32Rem(index, out int extraBits);
+
+                Debug.Assert(elementIndex < _array.Length, $"index: {index}, Length: {Length}, elementIndex: {elementIndex}, arrayLength: {_array.Length}, extraBits: {extraBits}");
+
+                int newValue = _array[elementIndex];
+                if (value)
+                {
+                    newValue |= 1 << extraBits;
+                }
+                else
+                {
+                    newValue &= ~(1 << extraBits);
+                }
+                _array[elementIndex] = newValue;
+            }
+        }
+
+        private void Grow(int index)
+        {
+            Debug.Assert(index >= Length, $"index: {index}, Length: {Length}");
+
+            // If index is a multiple of 32, add 1.
+            int newints = (index & 31) == 0 ? 1 : 0;
+
+            if ((uint)index / 2 < int.MaxValue)
+            {
+                index *= 2; // Grow by doubling, if possible.
+            }
+            else
+            {
+                index++;
```

Comment:
In this case, should we just set index to `int.MaxValue` instead of growing by 1? Growing by 1 could really slow things down in the unlikely but possible case of JSON with `depth > int.MaxValue/2`

Reply:
Always doubling the array length now (there is no risk of overflow since the integer array length is capped).